zokrates_core/src/static_analysis/constant_inliner.rs

```diff
@@ -0,0 +1,848 @@
+
+        inliner.fold_program(p)
+    }
+
+    fn module(&self) -> &TypedModule<'ast, T> {
+        self.modules.get(&self.location).unwrap()
+    }
+
+    fn change_location(&mut self, location: OwnedTypedModuleId) -> OwnedTypedModuleId {
+        let prev = self.location.clone();
+        self.location = location;
+        prev
+    }
+
+    fn get_constant(&mut self, id: &Identifier) -> Option<TypedConstant<'ast, T>> {
+        self.modules
+            .get(&self.location)
+            .unwrap()
+            .constants
+            .get(id.clone().try_into().unwrap())
+            .cloned()
+            .map(|symbol| self.get_canonical_constant(symbol))
+    }
+
+    fn get_canonical_constant(
+        &mut self,
+        symbol: TypedConstantSymbol<'ast, T>,
+    ) -> TypedConstant<'ast, T> {
+        match symbol {
+            TypedConstantSymbol::There(module_id, id) => {
+                let location = self.change_location(module_id);
+                let symbol = self.module().constants.get(id).cloned().unwrap();
+
+                let symbol = self.get_canonical_constant(symbol);
+                let _ = self.change_location(location);
+                symbol
+            }
+            TypedConstantSymbol::Here(tc) => TypedConstant {
+                expression: self.fold_expression(tc.expression),
+                ..tc
+            },
+        }
+    }
+}
+
+impl<'ast, T: Field> Folder<'ast, T> for ConstantInliner<'ast, T> {
+    fn fold_program(&mut self, p: TypedProgram<'ast, T>) -> TypedProgram<'ast, T> {
+        TypedProgram {
+            modules: p
+                .modules
+                .into_iter()
+                .map(|(module_id, module)| {
+                    self.change_location(module_id.clone());
+                    (module_id, self.fold_module(module))
+                })
+                .collect(),
+            main: p.main,
+        }
+    }
+
+    fn fold_module(&mut self, p: TypedModule<'ast, T>) -> TypedModule<'ast, T> {
+        TypedModule {
+            constants: p
+                .constants
+                .into_iter()
+                .map(|(key, tc)| (key, self.fold_constant_symbol(tc)))
+                .collect(),
+            functions: p
+                .functions
+                .into_iter()
+                .map(|(key, fun)| (key, self.fold_function_symbol(fun)))
+                .collect(),
+        }
+    }
+
+    fn fold_constant_symbol(
+        &mut self,
+        p: TypedConstantSymbol<'ast, T>,
+    ) -> TypedConstantSymbol<'ast, T> {
+        let tc = self.get_canonical_constant(p);
+        TypedConstantSymbol::Here(tc)
+    }
+
+    fn fold_field_expression(
+        &mut self,
+        e: FieldElementExpression<'ast, T>,
+    ) -> FieldElementExpression<'ast, T> {
+        match e {
+            FieldElementExpression::Identifier(ref id) => match self.get_constant(id) {
+                Some(c) => fold_field_expression(self, c.expression.try_into().unwrap()),
```

Comment:
It would be clearer to do something like

```
let c = self.fold_constant(c)
c.try_into().unwrap()
```

Using the `fold_constant` trait method suggested in the other comment

Reply:
if this `TryInto` isn't implemented, it would make sense to add it: downcast a constant to an expression of a given type. It can reuse the `TryInto` defined for `TypedExpression`, so that we don't access c.expression here. Ideally this struct field would be private and everything happens with methods.